File: HexChat/undo.py
from __future__ import print_function
from collections import deque
from sys import platform
import xchat as hexchat

__module_name__ = "Undo"
__module_author__ = "TingPing"
__module_version__ = "1"
__module_description__ = "Binds ctrl+z to undo and ctrl+y/shift+ctrl+z to redo."

undolevels = 10
redolevels = 10

undobufs = {}
redobufs = {}

if platform == 'darwin':
	primarymod = 1 << 28
else:
	primarymod = 1 << 2
shiftmod = 1 << 0

def get_valid_mod (mod):
	"""Modifiers are full of junk we dont care about, remove them"""
	return int(mod) & (1 << 0 | 1 << 2 | 1 << 3 | 1 << 28)

def keypress_cb(word, word_eol, userdata):
	global undobufs
	global redobufs
	bufname = '{}_{}'.format(hexchat.get_info('channel'), hexchat.get_info('network'))
	key = word[0]
	mod = get_valid_mod(word[1])
	inputtext = hexchat.get_info('inputbox')

	# Previous strings are stored as deque's in a dict for each channel
	if not bufname in undobufs:
		undobuflist = undobufs[bufname] = deque(maxlen=undolevels)
	else:
		undobuflist = undobufs[bufname]

	if not bufname in redobufs:
		redobuflist = redobufs[bufname] = deque(maxlen=redolevels)
	else:
		redobuflist = redobufs[bufname]

	if (key, mod) == ('122', primarymod): # ctrl+z
		try:
			# Get last saved string
			text = undobuflist.pop()
			if text == inputtext: # First undo may result in same text
				redobuflist.append(text)
				text = undobuflist.pop()
			hexchat.command('settext {}'.format(text))
			hexchat.command('setcursor {}'.format(len(text)))

			redobuflist.append(text)

		except IndexError: pass # No undos left

	elif ((key, mod) == ('121', primarymod) or # ctrl+y
			(key, mod) == ('90', shiftmod|primarymod)): # ctrl+shift+z 
		try:
			text = redobuflist.pop()
			if text == inputtext:
				text = redobuflist.pop()
			hexchat.command('settext {}'.format(text))
			hexchat.command('setcursor {}'.format(len(text)))

		except IndexError: pass

	else:
		# Just throw anything else in here if it has changed
		if not undobuflist or undobuflist[-1] != inputtext:
			undobuflist.append(inputtext)

def unload_cb(userdata):
	print(__module_name__, 'version',  __module_version__, 'unloaded.')

hexchat.hook_print('Key Press', keypress_cb) 
hexchat.hook_unload(unload_cb)
print(__module_name__, 'version',  __module_version__, 'loaded.')

File: HexChat/undo.lua
-- SPDX-License-Identifier: MIT
local bit = bit
if not bit then
	bit = require('bit32')
end
local lgi = require('lgi')
local Gdk = lgi.require('Gdk', '2.0')

hexchat.register('Undo', '1', 'Add keybindings to undo/redo')

local MAX_LEVELS = 10

local undobufs = {}
local redobufs = {}


local stack = {}

function stack:new ()
	local new_stack = {}
	setmetatable(new_stack, {__index=stack})
	return new_stack
end

function stack:push (val)
	if #self == MAX_LEVELS then
		table.remove(self, 1)
	end

	self[#self + 1] = val
end

function stack:pop ()
	local val = self[#self]
	self[#self] = nil
	return val
end


local function get_valid_mod (modifier)
	-- We only care about a few modifiers
	local mods = 0
	for _, mod in pairs({
		Gdk.ModifierType.CONTROL_MASK,
		Gdk.ModifierType.SHIFT_MASK,
		Gdk.ModifierType.MOD1_MASK, -- alt
	}) do
		mods = bit.bor(mods, mod)
	end

	return bit.band(tonumber(modifier), mods)
end

hexchat.hook_print('Key Press', function (args)
	local network = hexchat.get_info('network')
	if network == nil then
		network = '<none>'
	end
	local bufname = hexchat.get_info('channel') .. '_' .. network
	local key = tonumber(args[1])
	local mod = get_valid_mod(args[2])
	local input = hexchat.get_info('inputbox')

	local undostack = undobufs[bufname] or stack:new()
	local redostack = redobufs[bufname] or stack:new()
	if #undostack == 0 and #redostack == 0 then
		undobufs[bufname] = undostack
		redobufs[bufname] = redostack
	end

	if key == Gdk.KEY_z and mod == Gdk.ModifierType.CONTROL_MASK then
		local text = undostack:pop()
		if text then
			if text == input then
				redostack:push(text)
				text = undostack:pop() or ''
			end
			hexchat.command('settext ' .. text)
			hexchat.command('setcursor ' .. tostring(#text))
			redostack:push(text)
		end
	elseif (key == Gdk.KEY_y and mod == Gdk.ModifierType.CONTROL_MASK) or
	       (key == Gdk.KEY_Z and mod == bit.bor(Gdk.ModifierType.CONTROL_MASK, Gdk.ModifierType.SHIFT_MASK))  then
		local text = redostack:pop()
		if text then
			if text == input then
				text = redostack:pop() or ''
			end
			hexchat.command('settext ' .. text)
			hexchat.command('setcursor ' .. tostring(#text))
		end
	else -- Any key press
		if undostack[#undostack] ~= input then
			undostack:push(input)
		end
	end
end)
